rfc/plugins.md

```diff
@@ -0,0 +1,300 @@
+
+Initial plugin targets will be:
+
+* Pipeline Stages
+* Pipeline SpEL functions
+* Micrometer backends
+* Gate proxy endpoint configuration
+
+## Motivation and Rationale
+Extending Spinnaker today falls into two categories, 1) writing Java code that directly extends internal APIs, or 2) usage of web hook stages or preconfigured jobs.
+
+In the first case, a native Spring extension model, extensions need to be written on the JVM and released in lock-step with service releases. Further, it usually requires running the Spinnaker ecosystem to validate correct behavior - a technical hurdle that many possible contributors do not have the time or patience for. In the second case, extensions can be written and tested outside of the Spinnaker release cycle, but can only affect functionality of Spinnaker in the narrow scope of a pipeline stage.
+
+In an effort to lower the bar for Spinnaker development, a native plugin model is being proposed that can be used for federated development of Spinnaker. This native plugin system will need to support clearly defined extension points within Spinnaker at both service- and product-level abstractions, from either an in-process or remote runtime model.
+
+By creating this plugin model, Spinnaker will be able to iterate functionality faster as changes will be more easily developed, distributed and installed without adding bloat to the core product offering. This work aims to be a quality of life improvement for Spinnaker developers, operators, and users alike.
+
+Early stakeholders for this effort are:
+
+* Netflix Delivery Engineering
+* Netflix Cloud Database Engineering
+* Armory
+
+We have explicitly not called out the Spinnaker OSS community, as the early deliverables of this effort are experimental and won’t be subject to backwards compatibility constraints. Community members are welcome to be involved early nonetheless, provided acceptance of instability.
+
+## Timeline
+Proof of concepts were developed both by Armory and Netflix independently in 2019Q3. In Q4, we plan to deliver an alpha-quality plugin implementation that exhibits capability of extending multiple services as well as Deck, and include a basic distribution story for both in-process and out-of-process extensions.
+
+Development will continue into and through 2020. If the plugin proposal is successful, plugin development and thereby the core plugin framework will continue iteration for the foreseeable future.
+
+## Design
+### Taxonomies
+| Term | Definition |
+| - | - |
+| **Extension Point** | A strongly typed service- or library-defined point of extension |
+| **Plugin** | An implementation of an Extension Point |
+
+### Multiple Plugin Runtimes
+There’s a place and a time for different plugin runtimes. In-process extensibility offers a lot of power and network efficiency at the cost of higher coupling to core service rollouts, whereas a remote plugin runtime trades runtime efficiency and reliability for development and release efficiency.
```

Comment:
>  whereas a remote plugin runtime trades runtime efficiency and reliability for development and release efficiency.

A remote model also allows developer to pick their programming language/framework/runtime of choice.